library/core/src/cell.rs

```diff
@@ -1922,7 +1922,8 @@ impl<T: ?Sized + fmt::Display> fmt::Display for RefMut<'_, T> {
 ///
 /// - At all times, you must avoid data races. If multiple threads have access to
 /// the same `UnsafeCell`, then any writes must have a proper happens-before relation to all other
-/// accesses (or use atomics).
+/// accesses, or all accesses must use atomic operations that establish a memory barrier (e.g.,
+/// operations in [core::sync::atomic] or compiler intrinsics).
```

Comment:
"Establish a memory barrier" is not the correct term -- having happens-before is indeed more correct. Rust doesn't even have a concept of "memory barrier". We have memory *fences*, but one can establish happens-before without using fences.

I think the wording in https://github.com/rust-lang/rust/pull/128778 should fix your concern, we just have to wait for FCP to pass before it can land because I made it part of another PR that requires FCP. If you want, you can take just that change and put it here, then we can land it separately.

Reply:
I have updated https://github.com/rust-lang/rust/pull/128778 to avoid the pitfall with only talking about writes. I am open to suggestions for how to improve the wording!

Reply:
I think that the text added / changed in 3e5a4030b5c81b6698485bdbfc5d1f5d1be3613d is great, in particular:

> The most important aspect of this model is that *data races* are undefined behavior. A data race is defined as conflicting non-synchronized accesses where at least one of the accesses is non-atomic. Here, accesses are *conflicting* if they affect overlapping regions of memory and at least one of them is a write. They are *non-synchronized* if neither of them *happens-before* the other, according to the happens-before order of the memory model.

I'm worried that removing those concrete requirements from `UnsafeCell`'s documentation itself, and adding indirection by referring to `core::sync::atomic`, makes it harder to reason about `UnsafeCell`'s particular semantics though. I do see the appeal of de-duplicating this documentation though.

Reply:
I see, thanks for clarifying this! I agree that the current text in #128778 resolves this concern fully.